src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php

```diff
@@ -123,38 +131,60 @@ private function registerHandlers(ContainerBuilder $container)
+            }
         }
 
-        $handlersLocatorMapping = array();
-        foreach ($handlersByMessage as $message => $handlers) {
-            if (1 === \count($handlers)) {
-                $handlersLocatorMapping['handler.'.$message] = current($handlers);
-            } else {
-                $d = new Definition(ChainHandler::class, array($handlers));
-                $d->setPrivate(true);
-                $serviceId = hash('sha1', $message);
-                $definitions[$serviceId] = $d;
-                $handlersLocatorMapping['handler.'.$message] = new Reference($serviceId);
+        $handlersLocatorMappingByBus = array();
+        foreach ($buses as $bus) {
+            $handlersLocatorMappingByBus[$bus] = array();
+        }
+        foreach ($handlersByBusAndMessage as $bus => $handlersByMessage) {
+            foreach ($handlersByMessage as $message => $handlersIds) {
+                if (1 === \count($handlersIds)) {
+                    $handlersLocatorMappingByBus[$bus]['handler.'.$message] = new Reference(current($handlersIds));
+                } else {
+                    $chainHandler = new Definition(ChainHandler::class, array(array_map(function (string $handlerId): Reference {
+                        return new Reference($handlerId);
+                    }, $handlersIds)));
+                    $chainHandler->setPrivate(true);
+                    $serviceId = hash('sha1', $bus.$message);
+                    $definitions[$serviceId] = $chainHandler;
+                    $handlersLocatorMappingByBus[$bus]['handler.'.$message] = new Reference($serviceId);
+                }
             }
         }
         $container->addDefinitions($definitions);
 
-        $handlerResolver = $container->getDefinition('messenger.handler_resolver');
-        $handlerResolver->replaceArgument(0, ServiceLocatorTagPass::register($container, $handlersLocatorMapping));
+        foreach ($buses as $bus) {
+            $container->register($resolverName = "$bus.messenger.handler_resolver", ContainerHandlerLocator::class)
+                ->setArgument(0, ServiceLocatorTagPass::register($container, $handlersLocatorMappingByBus[$bus]))
+            ;
+            if ($container->has($callMessageHandlerId = "$bus.middleware.call_message_handler")) {
+                $container->getDefinition($callMessageHandlerId)
+                    ->replaceArgument(0, new Reference($resolverName))
+                ;
+            }
+        }
 
         if ($container->hasDefinition('console.command.messenger_debug')) {
-            $container->getDefinition('console.command.messenger_debug')
-                ->replaceArgument(0, array_map(function (array $handlers): array {
-                    return array_map('strval', $handlers);
-                }, $handlersByMessage));
+            $debugCommandMapping = $handlersByBusAndMessage;
+            foreach ($buses as $bus) {
+                if (!isset($debugCommandMapping[$bus])) {
```

Comment:
see previous suggestion, i think this should be checked for at runtime

Reply:
Added check at runtime in debug command